File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.java
class ProductOfNumbers {

    int[] nums;
    int size;

    public ProductOfNumbers() {
        nums = new int[40000];
        size = 0;
    }

    public void add(int num) {
        nums[size] = num;
        size++;
    }

    public int getProduct(int k) {
        int product = 1;

        for (int i = size - k; i < size; i++) {
            product *= nums[i];
        }

        return product;
    }
}

File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.cpp
class ProductOfNumbers {
public:
vector<int> prefixproduct;
    ProductOfNumbers() {
        prefixproduct.push_back(1);
        
    }
    
    void add(int num) {
        if(num ==0){
            prefixproduct.clear();
            prefixproduct.push_back(1);
            return;
        }
        prefixproduct.push_back(prefixproduct.back() * num);
        
    }
    
    int getProduct(int k) {
        int size = prefixproduct.size();
        if(size <= k){
            return 0;

        }
        return prefixproduct[size -1]/ prefixproduct[size - k- 1];
        
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */